### PR TITLE
[2.0.x] Allow M73 without SDSUPPORT

### PR DESCRIPTION
**Pull request by benlye**
* Allow LCD_SET_PROGRESS_MANUALLY to be enabled without SDSUPPORT
* Add T parameter to set the time elapsed

### CODE REVIEW

**Comment by benlye** [comment]
I think it's worth mentioning that I had to constrain the T parameter to 1sec less than 100 days because setting a value of 100 days or greater caused Marlin to reset.  This is, I think, a bug in the way time values with days are rendered on the LCD.

Also, with `DOGM_SD_PERCENT` enabled, any value with days displays incorrectly.  For example, 50% with a time of 1 day 12:34 displays as `50% 1d 12:3`.  A time with 10 days displays as `50%10d 12:3`.  I suspect the problem is in [https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/libs/duration_t.h#L151-L155](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/libs/duration_t.h#L151-L155).

Unless someone else gets there first I'll try to figure out what those problems are.

**Comment by dot-bob** [comment]
I wonder if we should also add reporting print percentage using the firmware capabilities protocol as well: [http://www.reprap.org/wiki/Firmware_Capabilities_Protocol](http://www.reprap.org/wiki/Firmware_Capabilities_Protocol)

**Comment by benlye** [comment]
Rebased to take #9202 into account.

**Comment by benlye** [comment]
Travis seems to have been broken by https://github.com/MarlinFirmware/Marlin/commit/acb1e313746c721c470e8e045bdd98446b4c58c7.

#9214

**Comment by benlye** [comment]
This isn't ready yet.  

I just realized that it will prevent a running printjob_timer from updating the screen.  In truth, I wasn't aware of the way printjob_timer worked, and it largely provides the functionality I was aiming for with the `M73 T` parameter.  

I think that it might still be useful though so I'll attempt to make it co-exist with printjob_timer by only taking updates if there is no timer running.

**Comment by benlye** [comment]
I've now added a separate option for manual control of the progress bar timer, `LCD_SET_TIMER_MANUALLY`.  It is dependent on `LCD_SET_PROGRESS_MANUALLY`.

This solves the problem I created where print_job_timer was being ignored if 
 `LCD_SET_PROGRESS_MANUALLY` was enabled but timer updates weren't sent from the host.

It works as expected for me.  With `LCD_SET_PROGRESS_MANUALLY` I can manually control the progress bar with `M73` and print_job_timer is in control of the timer.  With `LCD_SET_PROGRESS_MANUALLY` and `LCD_SET_TIMER_MANUALLY` enabled print_job_timer has no effect on the display and manual updates are done via `M73 T`.

It may not be terribly useful to have manual control over the timer, but it was interesting to use `M73 T` commands to find a couple of bugs (#9202, #9220).  There is one more lurking, where Marlin resets if the timer exceeds 100 days (unlikely, I know).

Removing the `SDSUPPORT` dependency for `LCD_SET_PROGRESS_MANUALLY` is much more interesting to me and if no one wants `LCD_SET_TIMER_MANUALLY` I'd be happy to do a new PR just to change that.

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -615,6 +612,12 @@
 #if ENABLED(DOGLCD)
   // Show SD percentage next to the progress bar
   //#define DOGM_SD_PERCENT
+	
+  // Add an 'M73' G-code to set the current percentage
+  //#define LCD_SET_PROGRESS_MANUALLY
```
Comment (thinkyhead) [inline Marlin/Configuration_adv.h]: The character LCD also displays a progress bar and elapsed time.

**Review by thinkyhead — CHANGES_REQUESTED** [review]
The way this feature is structured, it fully disables the use of the print job's built-in timer, even though the timer still runs. It would be better to move `elapsed` into the `Stopwatch` class and:

- Add methods to set and clear the `elapsed` value
- Modify `Stopwatch` so `duration` returns `elapsed` when it's non-zero, otherwise the usual value. Since `Stopwatch` uses `duration` internally, rename it `_duration` and make a new `duration` method (that calls `_duration`) for the rest of Marlin to use.

Also, these features should be enabled for character LCD too. The code is almost the same.

**Comment by benlye** [comment]
@thinkyhead, the changes you're asking for make sense, but are beyond what I can do at the moment.

I've re-scoped this PR to just do the one thing I really wanted - allow M73 without SDSUPPORT.  I've stashed the other changes so I can pick that up again another day.

**Review by thinkyhead — APPROVED** [review]

**Comment by Zer0t3ch** [comment]
Looks like this currently only functions with full graphics LCDs, but it should work with others. (I've had a functional progress bar with my "ZONESTAR_LCD" in the past)

**Comment by thinkyhead** [comment]
@Zer0t3ch — Do you have `LCD_PROGRESS_BAR` enabled? Try the current bugfix branch and see if there's a visible progress bar when that option is enabled, and check whether you can change it with `M73` when `LCD_SET_PROGRESS_MANUALLY` is enabled without `SDSUPPORT`.

**Comment by Zer0t3ch** [comment]
@thinkyhead I'm on the Bugfix-v2 branch, and it won't compile for me at all with both `LCD_PROGRESS_BAR` and `LCD_SET_PROGRESS_MANUALLY` enabled. I believe this is because the progress bar timing is undefined, the timing is still declared inside an `if SDSUPPORT` block.

[This is the code that I think needs moved](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/Configuration_adv.h#L587). I think just moving that block up to line ~270 should fix it.

**EDIT - I gave it a try, and I'm still getting issues at compile-time, but it's a different issue.

Here's the tail-end of my output from before I moved that block

```
In file included from /home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:115:0:
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h: In function 'void lcd_set_custom_characters(uint8_t)':
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h:432:51: error: 'progress' was not declared in this scope
                 createChar_P(LCD_STR_PROGRESS[i], progress[i]);
                                                   ^
In file included from /home/pi/marlin/build/sketch/src/lcd/../inc/MarlinConfigPre.h:27:0,
                 from /home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:23:
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h: In function 'void lcd_implementation_status_screen()':
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h:895:74: error: 'PROGRESS_BAR_MSG_TIME' was not declared in this scope
     if (progress_bar_percent > 2 && (ELAPSED(millis(), progress_bar_ms + PROGRESS_BAR_MSG_TIME) || !lcd_status_message[0]))
                                                                          ^
/home/pi/marlin/build/sketch/src/lcd/../inc/../core/macros.h:189:40: note: in definition of macro 'PENDING'
 #define PENDING(NOW,SOON) ((long)(NOW-(SOON))<0)
                                        ^
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h:895:38: note: in expansion of macro 'ELAPSED'
     if (progress_bar_percent > 2 && (ELAPSED(millis(), progress_bar_ms + PROGRESS_BAR_MSG_TIME) || !lcd_status_message[0]))
                                      ^
/home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp: In function 'void lcd_status_screen()':
/home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:680:41: error: 'PROGRESS_BAR_MSG_TIME' was not declared in this scope
       if (ELAPSED(ms, progress_bar_ms + PROGRESS_BAR_MSG_TIME + PROGRESS_BAR_BAR_TIME))
                                         ^
/home/pi/marlin/build/sketch/src/lcd/../inc/../core/macros.h:189:40: note: in definition of macro 'PENDING'
 #define PENDING(NOW,SOON) ((long)(NOW-(SOON))<0)
                                        ^
/home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:680:11: note: in expansion of macro 'ELAPSED'
       if (ELAPSED(ms, progress_bar_ms + PROGRESS_BAR_MSG_TIME + PROGRESS_BAR_BAR_TIME))
           ^
/home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:680:65: error: 'PROGRESS_BAR_BAR_TIME' was not declared in this scope
       if (ELAPSED(ms, progress_bar_ms + PROGRESS_BAR_MSG_TIME + PROGRESS_BAR_BAR_TIME))
                                                                 ^
/home/pi/marlin/build/sketch/src/lcd/../inc/../core/macros.h:189:40: note: in definition of macro 'PENDING'
 #define PENDING(NOW,SOON) ((long)(NOW-(SOON))<0)
                                        ^
/home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:680:11: note: in expansion of macro 'ELAPSED'
       if (ELAPSED(ms, progress_bar_ms + PROGRESS_BAR_MSG_TIME + PROGRESS_BAR_BAR_TIME))
           ^
Using library LiquidCrystal at version 1.0.7 in folder: /home/pi/marlin/arduino/libraries/LiquidCrystal
exit status 1
```

and this is after:

```
In file included from /home/pi/marlin/build/sketch/src/lcd/ultralcd.cpp:115:0:
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h: In function 'void lcd_set_custom_characters(uint8_t)':
/home/pi/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h:432:51: error: 'progress' was not declared in this scope
                 createChar_P(LCD_STR_PROGRESS[i], progress[i]);
                                                   ^
Using library LiquidCrystal at version 1.0.7 in folder: /home/pi/marlin/arduino/libraries/LiquidCrystal
exit status 1

```

Looking at the relevant code, it looks to me like `progress` *should* be properly declared, but my experience with CPP is thin at best.

**Comment by thinkyhead** [comment]
Your `Configuration_adv.h` is screwy if `PROGRESS_BAR_MSG_TIME` and `PROGRESS_BAR_BAR_TIME` aren't defined when `LCD_PROGRESS_BAR` is defined. It should look like this:
```cpp
  // Show a progress bar on HD44780 LCDs for SD printing
  //#define LCD_PROGRESS_BAR

  #if ENABLED(LCD_PROGRESS_BAR)
    // Amount of time (ms) to show the bar
    #define PROGRESS_BAR_BAR_TIME 2000
    // Amount of time (ms) to show the status message
    #define PROGRESS_BAR_MSG_TIME 3000
    // Amount of time (ms) to retain the status message (0=forever)
    #define PROGRESS_MSG_EXPIRE   0
    // Enable this to show messages for MSG_TIME then hide them
    //#define PROGRESS_MSG_ONCE
    // Add a menu item to test the progress bar:
    //#define LCD_PROGRESS_BAR_TEST
  #endif
```

**Comment by thinkyhead** [comment]
Aha, yes. You nailed it. That is now fixed.

**Comment by Zer0t3ch** [comment]
@thinkyhead Glad I could help but I'm still having an issue. Normally I wouldn't ask here, but I think it's relevant. 

Somehow, enabling `LCD_SET_PROGRESS_MANUALLY` (and `LCD_PROGRESS_BAR`, as sanity check fails without it) is causing compilation for me to fail, like this:

`bugfix-2.0.x` branch:  
```
In file included from /home/zac/marlin/build/sketch/src/lcd/ultralcd.cpp:115:0:
/home/zac/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h: In function 'void lcd_set_custom_characters(uint8_t)':
/home/zac/marlin/build/sketch/src/lcd/ultralcd_impl_HD44780.h:432:51: error: 'progress' was not declared in this scope
                 createChar_P(LCD_STR_PROGRESS[i], progress[i]);
                                                   ^
Using library LiquidCrystal at version 1.0.7 in folder: /home/zac/marlin/arduino/libraries/LiquidCrystal
exit status 1
```

`bugfix-1.1.x` branch:  
```
In file included from /home/zac/marlin/build/sketch/ultralcd.cpp:107:0:
/home/zac/marlin/build/sketch/ultralcd_impl_HD44780.h: In function 'void lcd_set_custom_characters(uint8_t)':
ultralcd_impl_HD44780.h:427: error: 'progress' was not declared in this scope
                 createChar_P(LCD_STR_PROGRESS[i], progress[i]);
                                                   ^
/home/zac/marlin/build/sketch/ultralcd.cpp: In function 'void lcd_status_screen()':
ultralcd.cpp:648: error: 'card' was not declared in this scope
       progress_bar_percent = card.percentDone();
                              ^
Using library LiquidCrystal at version 1.0.7 in folder: /home/zac/marlin/arduino/libraries/LiquidCrystal
exit status 1
```

On a sidenote, have a look at [this](https://github.com/MarlinFirmware/Marlin/commit/e65a853c59de71c5bb2d61dd15a6024f98060342#r27755423) suggestion I made in regards to your commit. (dunno if you get notifications for that stuff)

**Comment by Zer0t3ch** [comment]
http://paste.ubuntu.com/p/7yQ3tqQvxC

That diff fixes it. It seems the the block declaring the `progress`  array(?) was still inside an `if SDSUPPORT` block. Moving it outside compiles without issue, though I haven't deployed it or tested any prints, yet.

**Comment by thinkyhead** [comment]
Thanks! In terms of code organization, things are only going to get better and less scattered around.

**Comment by Zer0t3ch** [comment]
Yep, I can already tell. The 2.0 branch looks loads more organized in file structure, alone. I can't wait to see what the future holds for this project.

**Pull request merged**